core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -749,6 +698,77 @@ private void allocateUnassigned() {
         }
 
         /**
+         * Make a decision for allocating an unassigned shard.
+         */
+        private Tuple<Decision, ModelNode> allocateUnassignedShard(final ShardRouting shard, final Set<ModelNode> throttledNodes) {
+            assert !shard.assignedToNode() : "not an unassigned shard: " + shard;
+            if (allocation.deciders().canAllocate(shard, allocation).type() == Type.NO) {
+                // NO decision for allocating the shard, irrespective of any particular node, so exit early
+                return Tuple.tuple(null, null);
```

Comment:
I wonder if this method should return NO instead of null. 
This also makes the line
```
UnassignedInfo.AllocationStatus allocationStatus =
                            decision == null ? UnassignedInfo.AllocationStatus.DECIDERS_NO :
                                UnassignedInfo.AllocationStatus.fromDecision(decision);
```
obsolete in another place.